### PR TITLE
RDISCROWD-7554: Reset presented when configured

### PR DESCRIPTION
**Pull request by dchhabda**
* add a new project settings config `reset_presented_time`
* update cache with task cancel information adding new key `pybossa:task_cancelled:user:{0}:task:{1}`
* fix the bug with updating project setting attribute whose value to be updated is to be changed from `True` to `False`

### CODE REVIEW

File: pybossa/api/project.py [hunk 1]
```diff
@@ -140,8 +140,7 @@ def _update_object(self, obj):
 
     def _update_attribute(self, new, old):
         for key, value in old.info.items():
-            if not new.info.get(key):
-                new.info[key] = value
+            new.info.setdefault(key, value)
```
Comment (dchhabda) [inline pybossa/api/project.py:143]: this change is unrelated to reset_presented_time. it is to fix the bug where a project setting attribute whose value to be updated is to be changed from `True` to `False` cannot be done because the condition `if not new.info.get` turns out to be `True` due to which old value is retained.

File: pybossa/api/__init__.py [hunk 2]
```diff
@@ -211,8 +211,21 @@ def new_task(project_id, task_id=None):
...
-                    guard.extend_task_presented_timestamp_expiry(task, user_id_or_ip)
+                    # reset presented time for projects configured to reset it
+                    project = project_repo.get(project_id)
+                    if not project:
+                        return abort(400)
+                    reset_presented_time = project.info.get("reset_presented_time", False)
+                    if reset_presented_time:
+                        cancelled_task = guard.retrieve_cancelled_timestamp(task, user_id_or_ip)
+                        if cancelled_task:
+                            # user is returning to same task upon cancelling the task
```
Comment (ciaracian) [inline pybossa/api/__init__.py:222]: Does the lock on the task remain even if the user has canceled the task? If not, how can we assume that it is the same user?

**Review by ciaracian — APPROVED** [review]
LGTM! Just one quick question in the comments

File: pybossa/api/__init__.py (hunk 2 above)
Comment (dchhabda) [reply pybossa/api/__init__.py:222]: Its the soft lock that is present on the task upon cancel task. With soft lock, other users can obtain the lock and also if same user returns, the soft lock mechanism presents same task to that user first.
